File: test/data/cs/test06/StateState4.cs
public class StateState4 : BaseState
{
    private StateController _stateController;
    private IControllee _controllee;
    public StateState4(StateController stateController, IControllee controllee) : base(controllee)
    {
        _stateController = stateController;
        _controllee = controllee;
    }
    public override void Setup(bool resume, bool deepResume)
    {
        if(!deepResume)
        {
            SetupSubState(_stateController.InstanceOflong2, resume);
        }
    }
    public override BaseState? TransitEnoughData()
    {
        BaseState? currentSubState = CurrentSubState();
        if(currentSubState != null)
        {
            BaseState? nextState = currentSubState.TransitEnoughData();
            return TransitBySubState(nextState);
        }
        return null;
    }
    public override BaseState? TransitNewData()
    {
        BaseState? currentSubState = CurrentSubState();
        if(currentSubState != null)
        {
            BaseState? nextState = currentSubState.TransitNewData();
            return TransitBySubState(nextState);
        }
        return null;
    }
    public override string GetStateName()
    {
        BaseState? currentSubState = CurrentSubState();
        if(currentSubState == null)
        {
            return "State3.State4(end)";
        }
        return currentSubState.GetStateName();
    }
    public override BaseState? GetParent()
    {
        return _stateController.InstanceOfState3;
    }
    public override int GetStateID()
    {
        return 4;
    }
}

File: test/data/cpp/test06/StateState4.cpp
#include "StateState4.h"
StateState4::StateState4(StateController* stateController, IControllee* controllee) : BaseState(controllee)
{
    _stateController = stateController;
    _controllee = controllee;
}
StateState4::~StateState4()
{
}
void StateState4::Setup(bool resume, bool deepResume)
{
    if(!deepResume)
    {
        SetupSubState(_stateController->InstanceOflong2, resume);
    }
}
BaseState* StateState4::TransitEnoughData()
{
    BaseState* currentSubState = CurrentSubState();
    if(currentSubState != nullptr)
    {
        BaseState* nextState = currentSubState->TransitEnoughData();
        return TransitBySubState(nextState);
    }
    return nullptr;
}
BaseState* StateState4::TransitNewData()
{
    BaseState* currentSubState = CurrentSubState();
    if(currentSubState != nullptr)
    {
        BaseState* nextState = currentSubState->TransitNewData();
        return TransitBySubState(nextState);
    }
    return nullptr;
}
const char* StateState4::GetStateName()
{
    BaseState* currentSubState = CurrentSubState();
    if(currentSubState == nullptr)
    {
        return "State3.State4(end)";
    }
    return currentSubState->GetStateName();
}
BaseState* StateState4::GetParent()
{
    return _stateController->InstanceOfState3;
}
int StateState4::GetStateID()
{
    return 4;
}